File: debop4s-timeperiod/src/test/scala/com/github/debop4s/timeperiod/tests/calendars/BusinessCaseTest.scala
package com.github.debop4s.timeperiod.tests.calendars

import com.github.debop4s.timeperiod._
import com.github.debop4s.timeperiod.tests.AbstractTimePeriodTest
import com.github.debop4s.timeperiod.timerange._
import com.github.debop4s.timeperiod.utils.Times

/**
 * BusinessCaseTest
 * Created by debop on 2014. 2. 18.
 */
class BusinessCaseTest extends AbstractTimePeriodTest {

    test("time range calendar") {

        val now = Times.now

        (0 until 500).par.foreach(i => {
            val current = now + i.day
            val currentFiveSeconds = TimeRange(Times.trimToSecond(current, 15), Times.trimToSecond(current, 20))

            YearRange(current).hasInside(currentFiveSeconds) should equal(true)
            HalfyearRange(current).hasInside(currentFiveSeconds) should equal(true)
            QuarterRange(current).hasInside(currentFiveSeconds) should equal(true)
            MonthRange(current).hasInside(currentFiveSeconds) should equal(true)
            WeekRange(current).hasInside(currentFiveSeconds) should equal(true)
            DayRange(current).hasInside(currentFiveSeconds) should equal(true)
            HourRange(current).hasInside(currentFiveSeconds) should equal(true)
            MinuteRange(current).hasInside(currentFiveSeconds) should equal(true)

        })

        val anytime = TimeRange()

        YearRange().hasInside(anytime) should equal(false)
        HalfyearRange().hasInside(anytime) should equal(false)
        QuarterRange().hasInside(anytime) should equal(false)
        MonthRange().hasInside(anytime) should equal(false)
        WeekRange().hasInside(anytime) should equal(false)
        DayRange().hasInside(anytime) should equal(false)
        HourRange().hasInside(anytime) should equal(false)
        MinuteRange().hasInside(anytime) should equal(false)
    }

}

File: debop4s-timeperiod/src/test/java/debop4s/timeperiod/tests/calendars/BusinessCaseTest.java
package debop4s.timeperiod.tests.calendars;

import debop4s.core.JAction1;
import debop4s.core.parallels.JParallels;
import debop4s.timeperiod.TimeRange;
import debop4s.timeperiod.tests.TimePeriodTestBase;
import debop4s.timeperiod.timerange.*;
import debop4s.timeperiod.utils.Times;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;
import org.junit.Test;

import static org.fest.assertions.Assertions.assertThat;

@Slf4j
public class BusinessCaseTest extends TimePeriodTestBase {

    @Test
    public void timeRangeCalendarTimeRange() {
        final DateTime now = Times.now();

        JParallels.run(500, new JAction1<Integer>() {
            @Override
            public void perform(Integer i) {
                DateTime current = now.plusDays(i);
                TimeRange currentFiveSeconds = new TimeRange(Times.trimToSecond(current, 15), Times.trimToSecond(current, 20));

                assertThat(new YearRange(current).hasInside(currentFiveSeconds)).isTrue();
                assertThat(new HalfyearRange(current).hasInside(currentFiveSeconds)).isTrue();
                assertThat(new QuarterRange(current).hasInside(currentFiveSeconds)).isTrue();
                assertThat(new MonthRange(current).hasInside(currentFiveSeconds)).isTrue();
                assertThat(new WeekRange(current).hasInside(currentFiveSeconds)).isTrue();
                assertThat(new DayRange(current).hasInside(currentFiveSeconds)).isTrue();
                assertThat(new HourRange(current).hasInside(currentFiveSeconds)).isTrue();
                assertThat(new MinuteRange(current).hasInside(currentFiveSeconds)).isTrue();
            }
        });

        TimeRange anytime = new TimeRange();

        assertThat(new YearRange().hasInside(anytime)).isFalse();
        assertThat(new HalfyearRange().hasInside(anytime)).isFalse();
        assertThat(new QuarterRange().hasInside(anytime)).isFalse();
        assertThat(new MonthRange().hasInside(anytime)).isFalse();
        assertThat(new WeekRange().hasInside(anytime)).isFalse();
        assertThat(new DayRange().hasInside(anytime)).isFalse();
        assertThat(new HourRange().hasInside(anytime)).isFalse();
        assertThat(new MinuteRange().hasInside(anytime)).isFalse();
    }
}
